File: core/graph.go
package core

import "fmt"

// Graph holds the adjacency structure of the task DAG.
type Graph struct {
	// dependents maps a task ID to the IDs of tasks that depend on it.
	dependents map[string][]string
	// inDegree maps each task ID to its number of dependencies.
	inDegree map[string]int
	// order is a topologically sorted list of task IDs.
	order []string
}

// BuildGraph constructs a DAG from a Workflow and validates it has no cycles.
func BuildGraph(workflow *Workflow) (*Graph, error) {
	g := &Graph{
		dependents: make(map[string][]string, len(workflow.Tasks)),
		inDegree:   make(map[string]int, len(workflow.Tasks)),
	}

	for id := range workflow.Tasks {
		g.inDegree[id] = 0
		if _, exists := g.dependents[id]; !exists {
			g.dependents[id] = nil
		}
	}

	for id, task := range workflow.Tasks {
		for _, dependency := range task.DependsOn {
			if _, exists := workflow.Tasks[dependency]; !exists {
				return nil, fmt.Errorf("graph: task %s: unknown dependency %q", id, dependency)
			}
			g.dependents[dependency] = append(g.dependents[dependency], id)
			g.inDegree[id]++
		}
	}

	order, err := topologicalSort(g.inDegree, g.dependents)
	if err != nil {
		return nil, err
	}
	g.order = order
	return g, nil
}

// Dependents returns the IDs of tasks that directly depend on the given task.
func (g *Graph) Dependents(taskID string) []string {
	return g.dependents[taskID]
}

// TopologicalOrder returns tasks in an order where all dependencies come before their dependents.
func (g *Graph) TopologicalOrder() []string {
	return g.order
}

// DownstreamSet returns the set of task IDs that are transitively downstream
// of taskID, including taskID itself.
func (g *Graph) DownstreamSet(taskID string) map[string]bool {
	result := make(map[string]bool)
	g.collectDownstream(taskID, result)
	return result
}

func (g *Graph) collectDownstream(taskID string, visited map[string]bool) {
	if visited[taskID] {
		return
	}
	visited[taskID] = true
	for _, dependent := range g.dependents[taskID] {
		g.collectDownstream(dependent, visited)
	}
}

// topologicalSort uses Kahn's algorithm to compute the topological order and detect cycles.
func topologicalSort(inDegree map[string]int, dependents map[string][]string) ([]string, error) {
	degree := make(map[string]int, len(inDegree))
	for id, d := range inDegree {
		degree[id] = d
	}

	queue := make([]string, 0, len(degree))
	for id, d := range degree {
		if d == 0 {
			queue = append(queue, id)
		}
	}

	order := make([]string, 0, len(inDegree))
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		order = append(order, current)

		for _, dependent := range dependents[current] {
			degree[dependent]--
			if degree[dependent] == 0 {
				queue = append(queue, dependent)
			}
		}
	}

	if len(order) != len(inDegree) {
		return nil, fmt.Errorf("graph: workflow contains a dependency cycle")
	}
	return order, nil
}

File: stagehand/core/graph.py
from __future__ import annotations

from stagehand.core.workflow import Workflow


class Graph:
    """Holds the adjacency structure of the task DAG."""

    def __init__(
        self,
        dependents: dict[str, list[str]],
        in_degree: dict[str, int],
        order: list[str],
    ) -> None:
        self._dependents = dependents
        self._in_degree = in_degree
        self._order = order

    def dependents(self, task_id: str) -> list[str]:
        """Returns the IDs of tasks that directly depend on the given task."""
        return self._dependents.get(task_id, [])

    def topological_order(self) -> list[str]:
        """Returns tasks in an order where all dependencies come before their dependents."""
        return self._order

    def downstream_set(self, task_id: str) -> set[str]:
        """Returns the set of task IDs that are transitively downstream of task_id, including itself."""
        result: set[str] = set()
        self._collect_downstream(task_id, result)
        return result

    def _collect_downstream(self, task_id: str, visited: set[str]) -> None:
        if task_id in visited:
            return
        visited.add(task_id)
        for dependent in self._dependents.get(task_id, []):
            self._collect_downstream(dependent, visited)


def build_graph(workflow: Workflow) -> Graph:
    """Constructs a DAG from a Workflow and validates it has no cycles."""
    dependents: dict[str, list[str]] = {task_id: [] for task_id in workflow.tasks}
    in_degree: dict[str, int] = {task_id: 0 for task_id in workflow.tasks}

    for task_id, task in workflow.tasks.items():
        for dependency in task.depends_on:
            if dependency not in workflow.tasks:
                raise ValueError(f"graph: task {task_id}: unknown dependency {dependency!r}")
            dependents[dependency].append(task_id)
            in_degree[task_id] += 1

    order = _topological_sort(in_degree, dependents)
    return Graph(dependents, in_degree, order)


def _topological_sort(
    in_degree: dict[str, int],
    dependents: dict[str, list[str]],
) -> list[str]:
    """Kahn's algorithm — computes topological order and detects cycles."""
    degree = dict(in_degree)
    queue = [task_id for task_id, d in degree.items() if d == 0]
    order: list[str] = []

    while queue:
        current = queue.pop(0)
        order.append(current)
        for dependent in dependents.get(current, []):
            degree[dependent] -= 1
            if degree[dependent] == 0:
                queue.append(dependent)

    if len(order) != len(in_degree):
        raise ValueError("graph: workflow contains a dependency cycle")

    return order
